### PR TITLE
All spans of a trace share sampling state

### PR DESCRIPTION
**Pull request by tiffon**
## Which problem is this PR solving?

Lays the ground work for making sampling decisions based on span tags. These sampling decisions should affect all spans of a trace (within a given process) regardless of which span triggered the decision.

### Sampler API

Enhance the sampler API to be able to make sampling decisions on:

- Span creation
- Setting the operation name (aside from the initial value)
- Setting tags

Add `src/_flow/legacy_sampler_v1.js` to capture the legacy sampler API.

Redefine `src/_flow/sampler.js` to match the new sampler API.

Implement `src/samplers/v2/const_sampler.js` with the new API.

Implement `src/samplers/adapt_sampler.js` to wrap legacy samplers with the new API.

### Share sampled state within a trace

Share sampling state across all span contexts of the same trace.

Add `src/samplers/v2/sampling_state.js` to share sampling state among spans of the same trace.

Revise `src/span_context.js` to use `SamplingState` instead of `_flags` and `_samplingFinalized`. Update various classes that referred to these fields.

### CODE REVIEW

File: src/_flow/legacy_sampler_v1.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// @flow
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (vprithvi) [inline src/_flow/legacy_sampler_v1.js]: Should this be `2019`?

Comment (tiffon) [reply src/_flow/legacy_sampler_v1.js]: The file has been renamed.

Comment (tiffon) [reply src/_flow/legacy_sampler_v1.js]: Will update it.

**Comment by yurishkuro** [comment]
closing in favor of #377

**Pull request closed without merging**